QuantConnect.Polygon/PolygonDataDownloader.cs

```diff
@@ -74,35 +75,64 @@ public IEnumerable<BaseData> Get(DataDownloaderGetParameters parameters)
+                    true, false, DataNormalizationMode.Raw, tickType);
 
-                Task.Run(() => Parallel.ForEach(symbols, targetSymbol =>
-                {
-                    var historyRequest = new HistoryRequest(startUtc, endUtc, dataType, targetSymbol, resolution, exchangeHours, dataTimeZone,
-                        resolution, true, false, DataNormalizationMode.Raw, tickType);
-                    foreach (var data in _historyProvider.GetHistory(historyRequest))
-                    {
-                        dataQueue.Add(data);
-                    }
-                })).ContinueWith(_ =>
-                {
-                    dataQueue.CompleteAdding();
-                });
+                var historyData = _historyProvider.GetHistory(historyRequest);
 
-                foreach (var data in dataQueue.GetConsumingEnumerable())
+                if (historyData == null)
                 {
-                    yield return data;
+                    return null;
                 }
+
+                return GetBaseData(historyData);
             }
-            else
+        }
+
+        private IEnumerable<BaseData>? GetBaseData(Symbol symbol, DateTime startUtc, DateTime endUtc, Type dataType, Resolution resolution, SecurityExchangeHours exchangeHours, DateTimeZone dataTimeZone, TickType tickType)
+        {
+            var dataQueue = new BlockingCollection<BaseData>();
+            var symbols = GetOptions(symbol, startUtc, endUtc);
+
+            Task.Run(() => Parallel.ForEach(symbols, targetSymbol =>
             {
-                var historyRequest = new HistoryRequest(startUtc, endUtc, dataType, symbol, resolution, exchangeHours, dataTimeZone, resolution,
-                    true, false, DataNormalizationMode.Raw, tickType);
-                foreach (var data in _historyProvider.GetHistory(historyRequest))
+                var historyRequest = new HistoryRequest(startUtc, endUtc, dataType, targetSymbol, resolution, exchangeHours, dataTimeZone,
+                    resolution, true, false, DataNormalizationMode.Raw, tickType);
+
+                var history = _historyProvider.GetHistory(historyRequest);
+
+                if (history == null)
                 {
-                    yield return data;
+                    return;
                 }
+
+                foreach (var data in history)
+                {
+                    dataQueue.Add(data);
+                }
+            })).ContinueWith(_ =>
+            {
+                dataQueue.CompleteAdding();
+            });
+
+            // Validate: data is not null, we have gotten anything at least
+            if (!dataQueue.TryTake(out _, TimeSpan.FromSeconds(10)))
+            {
+                return null;
+            }
+
+            return GetBaseData(dataQueue.GetConsumingEnumerable());
+        }
+
+        private IEnumerable<BaseData> GetBaseData(IEnumerable<BaseData> baseData)
```

Comment:
You can remove this method, just return  `dataQueue.GetConsumingEnumerable()` and `historyData` in the methods above.

Reply:
Great. It is working 😋